application_sdk/common/credential_providers.py

```diff
@@ -0,0 +1,243 @@
+"""
+Credential provider implementation for various credential sources.
+
+This module provides a framework for fetching credentials from different sources
+such as direct credentials, AWS Secrets Manager, and extensibility for other
+credential sources.
+"""
+
+import json
+from typing import Any, Dict, Protocol, List
+import collections.abc
+
+from application_sdk.common.logger_adaptors import get_logger
+from application_sdk.common.utils import parse_credentials_extra
+import traceback
+
+logger = get_logger(__name__)
+
+
+class CredentialProvider(Protocol):
+    """Protocol defining the interface for credential providers."""
+
+    async def get_credentials(self, source_credentials: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Get the resolved credentials from the source credentials.
+
+        Args:
+            source_credentials (Dict[str, Any]): The source credentials containing
+                metadata necessary to resolve actual credentials.
+
+        Returns:
+            Dict[str, Any]: The resolved credentials.
+        
+        Raises:
+            ValueError: If credentials cannot be resolved.
+        """
+        ...
+
+
+class DirectCredentialProvider(CredentialProvider):
+    """Provider for directly supplied credentials."""
+
+    async def get_credentials(self, source_credentials: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Return the direct credentials as is.
+
+        Args:
+            source_credentials (Dict[str, Any]): The direct credentials.
+
+        Returns:
+            Dict[str, Any]: The same credentials.
+        """
+        return source_credentials
+
+
+class AWSSecretsManagerCredentialProvider(CredentialProvider):
+    """Provider for credentials from AWS Secrets Manager."""
+
+    async def get_credentials(self, source_credentials: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
too long for a method, can we break this down into one which overrides and is simpler + a few private utility methods?